File: 2020_11/Find the Smallest Divisor Given a Threshold/my.py
from math import ceil

class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        lower, upper = 1, max(nums)
        ans = []
        while lower < upper:
            mid = (lower + upper) >> 1
            res = 0
            for i in nums:
                res += ceil(i / mid)
            
            if res > threshold:
                lower = mid + 1
            else:
                upper = mid
                
        return lower

File: 2020_11/Find the Smallest Divisor Given a Threshold/my.go
func smallestDivisor(nums []int, threshold int) int {
    lower, upper := 1, max(nums)
    for lower < upper {
        mid := (lower + upper) >> 1
        res := getDivSum(nums, mid)
        if res > threshold {
            lower = mid + 1
        } else {
            upper = mid
        }
    }
    return lower
}

func max(nums []int) int {
    result := nums[0]
    for _, i := range nums {
        if result < i {
            result = i
        }
    }
    return result
}

func getDivSum(nums []int, mid int) int {
    res := 0
    for _, i := range nums {
        if i % mid == 0 {
            res += i / mid
        } else {
            res += (i / mid) + 1
        }
    }
    return res
}
